### PR TITLE
KEYCLOAK-9068: IDP-initiated-flow is not working with REDIRECT binding

### PR DESCRIPTION
**Pull request by rmartinc**
Patch for KEYCLOAK-9068.

The SAML binding-type is POST if `Assertion Consumer Service Redirect Binding URL` or general `Master SAML Processing URL` is configured. It is only REDIRECT is `Assertion Consumer Service Redirect Binding URL` is configured and the other two are not set (empty).

Previously it was half done and it always defaulted to POST no matter the configuration.

### CODE REVIEW

File: services/src/main/java/org/keycloak/protocol/saml/SamlService.java [hunk 1]
```diff
@@ -655,19 +655,47 @@ public Response idpInitiatedSSO(@PathParam("client") String clientUrlName, @Quer
...
+    /**
+     * Checks the client configuration to return the redirect URL and the binding type.
+     * POST is preferred, only if the SAML_ASSERTION_CONSUMER_URL_POST_ATTRIBUTE
+     * and management URL are empty REDIRECT is chosen.
+     *
+     * @param client Client to create client session for
+     * @return a two string array [redirectUrl, bindingType] or null if error
+     */
+    private String[] getUrlAndBindingForIdpInitiatedSso(ClientModel client) {
+        String redirectPost = client.getAttribute(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_POST_ATTRIBUTE);
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/SamlService.java]: Please rename to `aclPost` or similar, `redirect` has a different meaning in this context. (The same below for `redirectGet`)

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/IdpInitiatedLoginTest.java [hunk 2]
```diff
@@ -71,6 +69,70 @@ public void testIdpInitiatedLogin() {
...
+    @Test
+    public void testIdpInitiatedLoginRedirect() {
+        ClientRepresentation clientRep = adminClient.realm(REALM_NAME).clients().findByClientId(SAML_CLIENT_ID_SALES_POST).get(0);
+        String url = clientRep.getAttributes().get(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_POST_ATTRIBUTE);
+        adminClient.realm(REALM_NAME).clients().get(clientRep.getId())
+                .update(ClientBuilder.edit(clientRep)
+                        .attribute(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_POST_ATTRIBUTE, null)
+                        .attribute(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_REDIRECT_ATTRIBUTE, url)
+                        .build());
+        try {
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/IdpInitiatedLoginTest.java]: same as above

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/IdpInitiatedLoginTest.java [hunk 3]
```diff
@@ -71,6 +69,70 @@ public void testIdpInitiatedLogin() {
...
+    public void testIdpInitiatedLoginPostAdminUrl() {
+        ClientRepresentation clientRep = adminClient.realm(REALM_NAME).clients().findByClientId(SAML_CLIENT_ID_SALES_POST).get(0);
+        String url = clientRep.getAttributes().get(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_POST_ATTRIBUTE);
+        adminClient.realm(REALM_NAME).clients().get(clientRep.getId())
+                .update(ClientBuilder.edit(clientRep)
+                        .adminUrl(url)
+                        .attribute(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_POST_ATTRIBUTE, null)
+                        .attribute(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_REDIRECT_ATTRIBUTE, null)
+                        .build());
+        try {
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/IdpInitiatedLoginTest.java]: replace the lines 74-82 with
```java
        try (Closeable c = ClientAttributeUpdater.forClient(adminClient, REALM_NAME, SAML_CLIENT_ID_SALES_POST)
          .setAdminUrl(url)
          .setAttribute(SamlConfigAttributes.SAML_ASSERTION_CONSUMER_URL_POST_ATTRIBUTE, null)
          .setAttribute(SamlConfigAttributes.SAML_ASSERTION_CONSUMER_URL_REDIRECT_ATTRIBUTE, null)
          .update()) {

```
(add `ClientAttributeUpdater.setAdminUrl` method similarly to other ones in `ClientAttributeUpdater` class)
Then you can remove the `finally` part.

File: services/src/main/java/org/keycloak/protocol/saml/SamlService.java [hunk 4]
```diff
@@ -680,26 +708,33 @@ public Response idpInitiatedSSO(@PathParam("client") String clientUrlName, @Quer
...
-        if (bindingType.equals(SamlProtocol.SAML_REDIRECT_BINDING)) {
-            redirect = client.getAttribute(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_REDIRECT_ATTRIBUTE);
-        } else {
-            redirect = client.getAttribute(SamlProtocol.SAML_ASSERTION_CONSUMER_URL_POST_ATTRIBUTE);
-        }
-        if (redirect == null) {
-            redirect = client.getManagementUrl();
+        String[] bindingProperties = getUrlAndBindingForIdpInitiatedSso(client);
+        if (bindingProperties == null) {
+            throw new RuntimeException("SAML assertion consumer url not set up");
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/protocol/saml/SamlService.java]: No `RuntimeException`s please. Return eg. `null` in such case and treat this value in the caller methods appropriately (just as with other errors, usually error page is shown). `IdentityBrokerService.samlIdpInitiatedSSO` is also affected by this change and must handle the `null` return value accordingly.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thanks for the PR. Please see comments inline.

File: services/src/main/java/org/keycloak/protocol/saml/SamlService.java (hunk 1 above)
Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/saml/SamlService.java]: I called them postUrl and getUrl...

**Review by hmlnarik — APPROVED** [review]
Lgtm, thanks for the changes!
To be merged once tests pass

**Pull request merged**